File: packages/devtools/src/citations.ts
import { readFileSync } from "node:fs";
import path from "node:path";

import { z } from "zod";

const fixture = z.object({
  patterns: z.array(z.object({ name: z.string(), pattern: z.string() })),
});

const CASES = path.resolve(import.meta.dirname, "../../../contracts/citation/cases.json");

const read = (): z.infer<typeof fixture> => {
  const parsed: unknown = JSON.parse(readFileSync(CASES, "utf8"));
  return fixture.parse(parsed);
};

type Citation = { readonly what: string; readonly cited: string };

const patterns: readonly { readonly what: string; readonly pattern: RegExp }[] =
  read().patterns.map((one) => ({ what: one.name, pattern: new RegExp(one.pattern) }));

export const citationIn = (prose: string): Citation | undefined => {
  for (const { what, pattern } of patterns) {
    const found = pattern.exec(prose);
    if (found !== null) return { what, cited: found[0] };
  }
  return undefined;
};

File: packages/devtools/python/citations.py
from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any

CASES = Path(__file__).resolve().parents[3] / "contracts" / "citation" / "cases.json"


def _compiled() -> tuple[tuple[str, re.Pattern[str]], ...]:
    fixture: Any = json.loads(CASES.read_text(encoding="utf8"))
    read: list[tuple[str, re.Pattern[str]]] = []
    for one in fixture["patterns"]:
        what: str = one["name"]
        pattern: str = one["pattern"]
        read.append((what, re.compile(pattern)))
    return tuple(read)


_PATTERNS = _compiled()


def citation_in(prose: str) -> tuple[str, str] | None:
    for what, pattern in _PATTERNS:
        found = pattern.search(prose)
        if found is not None:
            return what, found.group(0)
    return None
